### PR TITLE
Addition of LinkCosigner and handling of eosjs transact parameters

### LINKED ISSUE #7: Implement OBFA for Fuel (and others) within the anchor-link signature collection process
Currently we have a conflict between:

- Signers using ESR (Anchor, etc)
- Wallet Providers (Anchor Link, UAL, Transit)

Which prevents signers from appending OBFA (only bill first authorizer) action onto signing requests from providers. 

Link, being used across all of these components, uses sessions which set the `broadcast: false` flag on every request passed between the provider and signer. Wallets detect this value being set to false, and refused to append OBFA actions like Fuel to the transaction being signed. 

What we need is likely to modify the link itself (being the core of the UAL and Transit plugins) to have the ability to collect a signature from an OBFA signature before pushing the transaction to the blockchain itself. 

The proposed new flow of a transaction in link would be something similar to:

1) Application passed transaction to Provider.
2) [New] Provider prepends OBFA action/authorization to transaction.
3) Provider uses open channel to pass a signing request to the Wallet.
4) Wallet signs transaction and issues callback, returning the transaction to Provider.
5) [New] Provider submits transaction to designated non-push API for the OBFA action to generate the signature of the OBFA provider.
6) [New] Provider combines signature into transaction
7) Provider submits transaction to the blockchain.

### PR DESCRIPTION
**Pull request by aaroncox**
Closes #7

### CODE REVIEW

File: src/link.ts [hunk 1]
```diff
@@ -298,32 +298,39 @@ export class Link implements esr.AbiProvider {
...
-                transaction: {
-                    expiration: '1970-01-01T00:00:00',
-                    ref_block_num: 0,
-                    ref_block_prefix: 0,
-                    max_net_usage_words: 0,
-                    max_cpu_usage_ms: 0,
-                    delay_sec: 0,
-                    ...anyArgs,
-                },
+        try {
```
Comment (jnordberg) [inline src/link.ts:305]: Why is this needed?

File: src/link.ts [hunk 2]
```diff
@@ -298,32 +298,39 @@ export class Link implements esr.AbiProvider {
...
+                        ...anyArgs,
+                    },
+                }
             }
+            const request = await this.createRequest(args, t)
+            const result = await this.sendRequest(request, t, broadcast)
+            return result
+        } catch (error) {
+            if (t.onFailure) {
+                t.onFailure(undefined, error)
```
Comment (jnordberg) [inline src/link.ts:334]: This will cause onFailure to be called multiple times with different arguments for the same error

File: src/link.ts (hunk 1 above)
Comment (aaroncox) [reply src/link.ts:305]: To catch and throw errors during `this.createRequest` mainly.

Comment (aaroncox) [reply src/link.ts]: Added a new commit that moved this into the createRequest call, which I think also addresses your other comment about it potentially being called multiple times?

Comment (aaroncox) [reply src/link.ts]: Also just for clarity, it allows our error messages to intercept things like this:

image

Comment (jnordberg) [reply src/link.ts]: That's not a very helpful error message to the user, not showing those errors was intentional, it's more of a "you're calling it wrong" than a "something unexpected went wrong with your transaction"

Comment (aaroncox) [reply src/link.ts]: They are not, but I'd rather get screenshots from users with this error message than having to tell them to open their browser console and give me the error reported there.

Comment (jnordberg) [reply src/link.ts]: Ok, guess it's not a huge deal now that the app can choose to hide error/success messages.

File: src/link-transport.ts [hunk 3]
```diff
@@ -19,7 +19,7 @@ export interface LinkTransport {
     /** Called if the request was successful. */
     onSuccess?(request: SigningRequest, result: TransactResult): void
     /** Called if the request failed. */
-    onFailure?(request: SigningRequest, error: Error): void
+    onFailure?(request: SigningRequest | undefined, error: Error): void
```
Comment (jnordberg) [inline src/link-transport.ts:22]: This is a API breaking change, if we want to do this now we should refactor all transport methods to take an identifier

Comment (aaroncox) [reply src/link-transport.ts:22]: I can look at this tomorrow unless you beat me to it. 

I'd like to get a release out tomorrow if possible and start updating all the wallet providers afterwards. You have any other must-have changes before that happens?

Comment (aaroncox) [reply src/link-transport.ts]: I went ahead and just reverted this single part of the code - while I still want this as a feature, I don't want it to hold up release.
